app/src/main/java/fr/neamar/kiss/dataprovider/simpleprovider/CalculatorProvider.java

```diff
@@ -1,73 +1,70 @@
-        // Now create matcher object.
-        Matcher m = p.matcher(query);
+        // Now create matcher objct.
+        Matcher m = P.matcher(query.replaceAll("\\s+", ""));
         if (m.find()) {
-            String operator = m.group(3);
+            String operation = m.group();
 
-            // let's go for floating point arithmetic
-            // we need to add a "0" on top of it to support ".2" => 0.2
-            // For every other case, this doesn't change the number "01" => 1
-            float lhs = Float.parseFloat("0" + m.group(2));
-            lhs = m.group(1).equals("-") ? -lhs : lhs;
-            float rhs = Float.parseFloat("0" + m.group(5));
-            rhs = m.group(4).equals("-") ? -rhs : rhs;
+            Result<ArrayDeque<Tokenizer.Token>> tokenized = Tokenizer.tokenize(operation);
+            String readableResult;
 
-            float floatResult = 0;
-            switch (operator) {
-                case "+":
-                    floatResult = lhs + rhs;
-                    break;
-                case "-":
-                    floatResult = lhs - rhs;
-                    break;
-                case "*":
-                case "×":
-                case "x":
-                    floatResult = lhs * rhs;
-                    operator = "×";
-                    break;
-                case "/":
-                case "÷":
-                    floatResult = lhs / rhs;
-                    operator = "÷";
-                    break;
-                default:
-                    floatResult = Float.POSITIVE_INFINITY;
+            if(tokenized.syntacticalError) {
+                return;
+            } else if(tokenized.arithmeticalError) {
+                return;
+            } else {
+                Result<ArrayDeque<Tokenizer.Token>> posfixed = ShuntingYard.infixToPostfix(tokenized.result);
+
+                if (posfixed.syntacticalError) {
+                    return;
+                } else if (posfixed.arithmeticalError) {
+                    return;
+                } else {
+                    Result<BigDecimal> result = Calculator.calculateExpression(posfixed.result);
+
+                    if (result.syntacticalError) {
+                        return;
+                    } else if (result.arithmeticalError) {
+                        return;
+                    } else {
+                        String localizedNumber = LOCALIZED_NUMBER_FORMATTER.format(result.result);
+                        readableResult = " = " + localizedNumber;
+                    }
+                }
             }
 
-            String queryProcessed = floatToString(lhs) + " " + operator + " "
-                    + floatToString(rhs) + " = " + floatToString(floatResult);
+            String queryProcessed = operation + readableResult;
             SearchPojo pojo = new SearchPojo("calculator://", queryProcessed, "", SearchPojo.CALCULATOR_QUERY);
 
             pojo.relevance = 100;
```

Comment:
That's not from you, but I believe this should now be 19. Otherwise, when you type a phone number, the math result appears before the call option

Reply:
The relevance? I have no idea what that means, and it is not related to this PR.

Reply:
It's the importance of a result -- how high in the list it should be. It's fine if you don't change it, I'll commit something after merging